core/src/main/java/hudson/model/AbstractCIBase.java

```diff
@@ -137,15 +135,20 @@ private void updateComputer(Node n, Map<String,Computer> byNameMap, Set<Computer
         used.add(c);
     }
 
-    /*package*/ void removeComputer(Computer computer) {
-        Map<Node,Computer> computers = getComputerMap();
-        for (Map.Entry<Node, Computer> e : computers.entrySet()) {
-            if (e.getValue() == computer) {
-                computers.remove(e.getKey());
-                computer.onRemoved();
-                return;
+    /*package*/ void removeComputer(final Computer computer) {
+        Queue.withLock(new Runnable() {
```

Comment:
Why not use a lambda ?
Oh sorry, JDK8 not yet approved as jenkins minimal runtime :P